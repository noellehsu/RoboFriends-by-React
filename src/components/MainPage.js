import React from 'react';
import SearchBox from '../components/SearchBox';
import CardList from '../components/CardList';
import Scroll from '../components/Scroll';
import ErrorBoundary from '../components/ErrorBoundary';
import './MainPage.css'
import Header from '../components/Header';


class MainPage extends React.Component {
    componentDidMount() {
        this.props.onRequestRobots();
        // fetch('https://jsonplaceholder.typicode.com/users')
        //     .then(response => response.json())
        //     .then(users => this.setState({ robots: users }));
    }

    //原本從onSearchChange(event)改成 onSearchChange = (event) =>，是為了確保裡面的this指向App，不然原本是指向SearchBox的<input>
    //onSearchChange在上面已經被寫成屬性，這邊的函式註解掉
    // onSearchChange = (event) => {
    //     //不要寫成this.state.searchfield，直接用內建方法setState
    //     this.setState({ searchfield: event.target.value });
    // }

    filterRobots = () =>{
       return this.props.robots.filter(robot => {
            return robot.name.toLowerCase().includes(this.props.searchField.toLowerCase());
        })
    }

    render() {
        const {onSearchChange, isPending} = this.props;
      
        if (isPending) {
            return <h1 className='tc'>Loading</h1>
        } else {
            return (
                <div className='tc'>
                    <Header />
                    <SearchBox searchChange={onSearchChange} />
                    <Scroll>
                        <ErrorBoundary>
                            <CardList robots={this.filterRobots} />
                        </ErrorBoundary>
                    </Scroll>
                </div>
            );
        }


    }
}

//connect方法會回傳一個方法
// export default connect(mapStateToProps,mapDispatchToProps)(App);

export default MainPage;
