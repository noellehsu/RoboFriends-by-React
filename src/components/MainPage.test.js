import React from 'react';
import MainPage from './MainPage';
// import { shallow, mount, render } from 'enzyme';
import { shallow } from 'enzyme';

// setup file
import { configure } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';


configure({ adapter: new Adapter() })

// it('expect to render MainPage component', () => {
//     const mockStore = {
//         robot:[],
//         searchField:''
//     };
//     expect(shallow(<MainPage store={mockStore}/>)).toMatchSnapshot();

// })

let wrapper;
beforeEach(() => {
    // 在所有測試前都要先跑過這邊
    const mockProps = {
        onRequestRobots: jest.fn(),
        robot: [],
        searchField: '',
        isPending: false
    }

    wrapper = shallow(<MainPage {...mockProps} />)
})


it('render MainPage', () => {
    expect(wrapper).toMatchSnapshot();
})


it('filters robots correctly', () => {
    const mockProps2 = {
        onRequestRobots: jest.fn(),
        robot: [{
            id:3,
            name:'John',
            eami:'jo@gmail.com'
        }],
        searchField: 'john',
        isPending: false
    }
    const wrapper2 = shallow(<MainPage {...mockProps2} />)

    expect(wrapper2.instance().filterRobots([])).toEqual([{
        id:3,
        name:'John',
        eami:'jo@gmail.com'
    }])

})

it('filters robots correctly again', () => {
    const mockProps3 = {
        onRequestRobots: jest.fn(),
        robot: [{
            id:3,
            name:'John',
            eami:'jo@gmail.com'
        }],
        searchField: 'john',
        isPending: false
    }

    
    const wrapper3 = shallow(<MainPage {...mockProps3} />)

    expect(wrapper3.instance().filterRobots([])).toEqual([{
        id:3,
        name:'John',
        eami:'jo@gmail.com'
    }])

})


