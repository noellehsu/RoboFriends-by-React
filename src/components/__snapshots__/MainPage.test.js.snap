// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`filters robots correctly 1`] = `undefined`;

exports[`render MainPage 1`] = `ShallowWrapper {}`;
